drivers: rtc: Wrong alarm mask in rtc_alarm_get_time

Kampi opened the pull request:
Fixing the following issues:

1. The mask can be wrong if the alarm register is
set to a MAX value because the alarm bit is the
highest in the alarm register. The mask is now
generated by checking the AE_x bits in the time
registers.

2. Fixing possible NULL pointer exception in
alarm_set_time API. timeptr can be set to NULL
with mask 0 in the alarm_set_time function.
The regs variable for the I2C communication
is written with the correct value from timeptr
only when the right bit in the mask is set.

3. rv8263c8_alarm_set_time() now resets the
alarm status.

4. Interrupts are now enabled by using
rv8263c8_alarm_set_time() rather than when
setting an alarm callback.

Closes #76659

Kampi committed:
drivers: rtc: Wrong alarm mask in rtc_alarm_get_time

Fixing 3 issues:

1. The mask can be wrong if the alarm register is
set to a MAX value because the alarm bit is the
highest in the alarm register. The mask is now
generated by checking the AE_x bits in the time
registers.

2. Fixing possible NULL pointer exception in
alarm_set_time API. timeptr can be set to NULL
with mask 0 in the alarm_set_time function.
The regs variable for the I2C communication
is written with the correct value from timeptr
only when the right bit in the mask is set.

3. rv8263c8_alarm_set_time() now resets the
alarm status.

4. Interrupts are now enabled by using
rv8263c8_alarm_set_time() rather than when
setting an alarm callback.

Signed-off-by: Daniel Kampert <DanielKampert@kampis-elektroecke.de>

Kampi commented:
@jakubtopic please check it :)

fabiobaltieri commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -133,6 +133,24 @@ static int rv8263c8_update_disable_timer(const struct device *dev)
+{
+	const struct rv8263c8_config *config = dev->config;
+
+	return i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
+				      RV8263C8_BM_ALARM_INT_ENABLE, RV8263C8_BM_ALARM_INT_DISABLE) |
+	       i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_SECONDS_ALARM,
+				     RV8263C8_BM_ALARM_DISABLE) |
+	       i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_MINUTES_ALARM,
+				     RV8263C8_BM_ALARM_DISABLE) |
+	       i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_HOURS_ALARM,
+				     RV8263C8_BM_ALARM_DISABLE) |
+	       i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_DATE_ALARM,
+				     RV8263C8_BM_ALARM_DISABLE) |
+	       i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_WEEKDAY_ALARM,
+				     RV8263C8_BM_ALARM_DISABLE);
```
weird... do the calls one at a time and check for each one, no sense cutting this short and letting a failed transaction fall back to the next one, also the indentation looks odd

fabiobaltieri commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -529,27 +545,28 @@ static int rv8263c8_alarm_get_time(const struct device *dev, uint16_t id, uint16
 
-	if (value[0] <= MAX_SEC) {
+	/* Check if the highest bit is not set. If so the alarm is enabled. */
+	if (value[0] & (~RV8263C8_BM_ALARM_ENABLE)) {
```
I think the parenthesis are not needed here

Kampi commented:
@fabiobaltieri I changed the logic for the `rv8263c8_alarm_get_time` functions because the logic was wrong.

jakubtopic commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -424,9 +443,7 @@ static int rv8263c8_alarm_set_time(const struct device *dev, uint16_t id, uint16
 	/* Disable the alarm when mask is zero. */
 	if (mask == 0) {
-		return i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
-					      RV8263C8_BM_ALARM_INT_ENABLE,
-					      RV8263C8_BM_ALARM_INT_DISABLE);
+		return rv8263c8_disable_alarm(dev);
 	}
```
This leaves interrupts incorrectly disabled after disabling and re-enabling the alarm using `rtc_alarm_set_time()` (if the interrupts were previously enabled).

I think you can remove these lines entirely, as setting the AE_x bits is handled correctly by the following code.

Kampi replied:
You are right. If `mask` is set to 0 no alarm will be enabled.

Kampi replied:
But I think I can use it to globally enable or disable the interrupts by setting or clearing the bit in the `CONTROL_2` register. Something like this

```
if (mask == 0) {
i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
				    RV8263C8_BM_ALARM_INT_ENABLE, RV8263C8_BM_ALARM_INT_DISABLE);
} else {
i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
				    RV8263C8_BM_ALARM_INT_ENABLE, RV8263C8_BM_ALARM_INT_ENABLE);
}
````

What do you think?

jakubtopic replied:
Sounds good to me :)

Kampi replied:
Ok then I would remove the alarm enabling and disabling completely from the `alarm_set_callback` API and move it to the `arlarm_set_time` API. So the alarm can only be disabled by calling `set_time` and `set_callback` can only remove the callback.

jakubtopic replied:
Yes! That's certainly feasible, solves the issue above and behaves according to the documentation.

jakubtopic commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -585,9 +603,7 @@ static int rv8263c8_alarm_set_callback(const struct device *dev, uint16_t id,
-		err = i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
-					     RV8263C8_BM_ALARM_INT_ENABLE,
-					     RV8263C8_BM_ALARM_INT_DISABLE);
+		err = rv8263c8_disable_alarm(dev);
```
`rtc_alarm_set_callback` should control only the alarm callback functionality. When disabling the alarm callback, the alarm should remain enabled until disabling it using `rtc_alarm_set_time()`.

Kampi replied:
However, the alarm should also not be enabled when setting the callback. The documentation is talking about the alarm callback and not the alarm itself.

jakubtopic replied:
There is definitely lack of documentation and inconsistent implementation regarding the interrupt outputs (I'll open a separate issue on this).

Regardless, calling `rv8263c8_disable_alarm()` here will disable both the alarm and alarm callback, but the documentation of `rtc_alarm_set_callback()` states, that the alarm should remain enabled and only be disabled by calling `rtc_alarm_set_time()`:

> Setting the alarm callback for an alarm, will enable the alarm callback. When the callback for an alarm is enabled, the alarm triggered event will invoke the callback, after which the alarm pending status will be cleared automatically. The alarm will remain enabled until manually disabled using [rtc_alarm_set_time()](https://docs.zephyrproject.org/apidoc/latest/group__rtc__interface.html#ga1fc9f0ec9642dfbf742ea541a20bdad5).
>
> To disable the alarm callback for an alarm, the callback and user_data parameters must be set to NULL. When the alarm callback for an alarm is disabled, the alarm triggered event will set the alarm status to "pending". To check if the alarm status is "pending", use [rtc_alarm_is_pending()](https://docs.zephyrproject.org/apidoc/latest/group__rtc__interface.html#ga531c379c437a91df44a45e95063700e8).

jakubtopic reviewed (changes requested):
Determining the alarm mask looks good! :)

I think there is an issue with usage of `rv8263c8_disable_alarm()` which combines both disabling the alarm and interrupt flags:

- In `rtc_alarm_set_time()`, you only really need to configure the alarm enable bits (AE_x). Disabling the AIE bit has no effect here. If you do disable it, you will need to re-enable it when the alarm is enabled again when calling `rtc_alarm_set_time()`.
- On contrary, [`rtc_alarm_set_callback()`](https://docs.zephyrproject.org/apidoc/latest/group__rtc__interface.html#ga77ccb9e6bd32e97016e01729d95074b6) should not affect the alarm enable bits, as it prevents checking the alarm status using `rtc_alarm_is_pending()` after the alarm callback is disabled.

Kampi commented:
I updated the drivers with the points from the discussion. I also found another possible bug. For `rtc_alarm_set_time()` the documentation says:

```
The timeptr param may be NULL if the mask param is 0
```

This can lead to a NULL pointer exception, because `regs` is filled with the values from `timeptr` which can be `NULL`. I moved the instructions to the conditional statements of each `mask` bit.

jakubtopic reviewed (commented):
Nice work, thanks. I have one suggestion. When filling `regs`, could you also set the AE_x bit and remove the individual i2c calls? This way, the registers can be written in a single access.

Kampi commented:
@bjarki-andreasen and @jakubtopic If I see it correctly the scenario with mask = 0 and timeptr = NULL isn't covered in the unit test. Otherwise this issue has been seen earlier.

jakubtopic commented:
Hm... The RTC test suite includes clearing the alarm time, so there is at least one call made with mask = 0 and timeptr = NULL. It likely wouldn't catch a potential null dereference, though.

In the `rtc_alarm_set_time()` in this driver, the null dereference issue wasn't originally present, until I suggested removing the condition checking if the mask is 0 :))

The other two mentioned issues could be covered by tests:
- Ensuring the interrupt output remains enabled after disabling and re-enabling the alarm using `rtc_alarm_set_time()`
- Verifying that `rtc_alarm_set_callback()` doesn’t disable the alarm unintentionally.

jakubtopic commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -422,89 +422,59 @@ static int rv8263c8_alarm_set_time(const struct device *dev, uint16_t id, uint16
 	}
 
-	return err;
+	return i2c_write_dt(&config->i2c_bus, regs, sizeof(regs));
```
brilliant 🙂

jakubtopic commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -422,89 +422,59 @@ static int rv8263c8_alarm_set_time(const struct device *dev, uint16_t id, uint16
 	} else {
-		err = i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_SECONDS_ALARM,
-					    RV8263C8_BM_ALARM_DISABLE);
+		err = i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
+					     RV8263C8_BM_ALARM_INT_ENABLE,
+					     RV8263C8_BM_ALARM_INT_ENABLE);
```
I now realized that enabling alarm interrupts prior to setting the alarm time could lead to an inadvertent alarm event.

Safer procedure for setting the alarm time would be:
1. always clear the AIE and AF flag
2. write the alarm settings
3. set the AIE to 1

Kampi replied:
Okay, I changed it. Now the behaviour of the function should be:

`mask` == 0:
- The `AIE` bit is cleared and alarms are globally disabled
- The `AE_x` bits are set and written to the alarm time registers to disable the time-based alarms
- `timeptr` is ignored completely

`mask` != 0:
- The `AIE` and `AIF` bit is cleared and alarms are globally disabled
- The `AE_x` bits are cleared and together written with the time from `timeptr` to the alarm time registers to enable the time-based alarms
- The `AIE` set and alarms are globally enabled

jakubtopic commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -529,27 +512,28 @@ static int rv8263c8_alarm_get_time(const struct device *dev, uint16_t id, uint16
 		return err;
 	}
 
-	if (value[0] <= MAX_SEC) {
```
Please also remove the `MAX_x` and `MIN_x` defines. They are no longer used and could misleading.

jakubtopic commented on drivers/rtc/rtc_rv8263.c:
```diff
@@ -422,86 +422,69 @@ static int rv8263c8_alarm_set_time(const struct device *dev, uint16_t id, uint16
-	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
-		err = i2c_reg_write_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_SECONDS_ALARM,
-					    RV8263C8_BM_ALARM_ENABLE | regs[0]);
+	if (mask == 0) {
+		err = i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
+					     RV8263C8_BM_ALARM_INT_ENABLE,
+					     RV8263C8_BM_ALARM_INT_DISABLE);
```
```suggestion
	if (mask == 0) {
		err = i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2,
					     RV8263C8_BM_ALARM_INT_ENABLE | RV8263C8_BM_AF,
					     RV8263C8_BM_ALARM_INT_DISABLE);
```
I would also clear the AF flag here.

jakubtopic reviewed (commented):
Looks good :)

Just a small suggestion for the commit message:
> 1. The mask can be wrong if the alarm register is
set to a MAX value because the alarm bit is the
highest in the alarm register.

Please mention resolving the issue by checking the `AE_x` flags

> 2. Disabling the alarm callback must also disable
the alarm flags in the alarm time registers
because otherwise rtc_alarm_get_time will
return a wrong mask.

_2._ can be dropped as it turns out it wasn't an issue.

Additionally, please document the changes in behaviour:
- `rv8263c8_alarm_set_time()` now resets the alarm status.
- RV8263's alarm interrupt output is now enabled by enabling the alarm using `rv8263c8_alarm_set_time()` rather than when setting an alarm callback.

jakubtopic commented:
@bjarki-andreasen, could you please review the updated behaviour of `rv8263c8_alarm_set_time()`? Specifically how it controls the RTC's interrupt output. This version seems better to me, as it supports utilizing the interrupt output even if it isn't connected to the MCU's GPIO.

bjarki-andreasen commented:
> @bjarki-andreasen, could you please review the updated behaviour of rv8263c8_alarm_set_time()? Specifically how it controls the RTC's interrupt output. This version seems better to me, as it supports utilizing the interrupt output even if it isn't connected to the MCU's GPIO.

I don't quite get how the interrupt output can be utilized if it is not connected :)

Since the alarm and update flags/events both use the same interrupt, the interrupt should always be enabled if either update or alarm callback is set. When either update or alarm trigger the interrupt, the driver should read the status (RV8263C8_REGISTER_CONTROL_2) register, check what actually triggered the interrupt (RV8263C8_BM_TF set and/or RV8263C8_BM_AF set), then clear the flags, then call the update and/or alarm callbacks if set. Lastly, if the alarm callback was not set, we must store the pending status of the alarm in `struct rv8263c8_data` to be read and cleared by a call to `rv8263c8_alarm_is_pending()`

Summarized, there should only be one work item for the one interrupt callback, which checks the status and handles both update and alarm events.

jakubtopic commented:
Sorry for changing the direction of this PR, no idea how we ended up here.

> I don't quite get how the interrupt output can be utilized if it is not connected :)

The scenario in question involves a use case, where the RTC's interrupt output is not connected to the MCU but to another component, like a power management IC. In this case interrupt gpios on the RTC node would not be present and if the `set_alarm_callback()` controls which events generate an interrupt signal, there would be no way to enable it.

> If the alarm callback was not set, we don't need to clear RV8263C8_BM_AF since it can't trigger the interrupt, and we instead read/clear it when the is_pending API is called :)

I believe the correct behaviour is to (also) clear the alarm status when setting a new alarm time. Otherwise if an alarm is already pending, the newly set alarm may never trigger.

> Summarized, there should only be one work item for the one interrupt callback, which checks the status and handles both update and alarm events.

No changes there.

Kampi commented:
> Sorry for changing the direction of this PR, no idea how we ended up here.
> 
> > I don't quite get how the interrupt output can be utilized if it is not connected :)
> 
> The scenario in question involves a use case, where the RTC's interrupt output is not connected to the MCU but to another component, like a power management IC. In this case interrupt gpios on the RTC node would not be present and if the `set_alarm_callback()` controls which events generate an interrupt signal, there would be no way to enable it.
> 
> > If the alarm callback was not set, we don't need to clear RV8263C8_BM_AF since it can't trigger the interrupt, and we instead read/clear it when the is_pending API is called :)
> 
> I believe the correct behaviour is to (also) clear the alarm status when setting a new alarm time. Otherwise if an alarm is already pending, the newly set alarm may never trigger.
> 
> > Summarized, there should only be one work item for the one interrupt callback, which checks the status and handles both update and alarm events.
> 
> No changes there.

If I remember correctly @bjarki-andreasen and I have discussed this question in the initial PR for the RTC. Theoretically, you can handle the alarm and update stuff by using a thread or a worker to poll the status flags and call the callbacks from there. But IMO it´s not a good approach and we decided to discard this idea.

Let's move this question into a discussion. IMO It´s a valid use case and when working with a PMIC for low-power applications you don´t have a thread or worker to poll the status flags constantly. So there should be a more generic idea.

jakubtopic commented:
> Let's move this question into a discussion. IMO It´s a valid use case and when working with a PMIC for low-power applications you don´t have a thread or worker to poll the status flags constantly. So there should be a more generic idea.

I agree with moving the discussion. Conclusion for this PR is that it turns on interrupt output when setting an alarm, which supports the use case I mentioned and is also consistent with some existing drivers.

bjarki-andreasen commented:
Now I'm catched up as well :) Regarding the interrupt, if it is not connected to the host, It makes sense to enable it (it has no negative impact) The reason for clearing the interrupt is to stop the IRQ from being called, which of course is only neccessary if the interrupt is connected :)

I'm okay with adding a note to the rtc_set_alarm_time which states the interrupt always is enabled if it is not physically connected to the host :)

Kampi commented:
@bjarki-andreasen and @fabiobaltieri can you review it, please?

bjarki-andreasen reviewed (approved):
Looks good :)
